tests/http-adapter.spec.js

```diff
@@ -1,198 +1,218 @@
-        .catch((err) => {
-          expect(err).toBeInstanceOf(Error)
-          expect(err.message).toEqual(expect.stringContaining(path))
-          expect(err.message).toEqual(expect.stringContaining(`${errorCode}`))
-        })
+      fetchMock.mock(/.*api\/user\/42\//g, errorCode, {
+        method: 'GET',
+        status: errorCode,
+        statusText: ''
+      })
+
+      try {
+        await adapter.get(path)
+      } catch (err) {
+        expect(err).toBeInstanceOf(Error)
+        expect(err.message).toEqual(expect.stringContaining(path))
+        expect(err.message).toEqual(expect.stringContaining(`${errorCode}`))
+      }
     })
   })
 
-  describe('#create', function () {
-    it('creates a new resource on the network', function () {
+  describe('#create', () => {
+    it('creates a new resource on the network', async () => {
       let payload = {data: {foo: 'bar', fiz: {biz: 'buz'}}}
-      server.respondWith('POST', '/api/user/', JSON.stringify(payload))
+      fetchMock.mock(/.*api\/user\//g, payload, {
+        method: 'POST'
+      })
 
-      return adapter.create('/api/user/', payload)
-        .then((response) => {
-          expect(response).toEqual(payload)
-        })
+      const response = await adapter.create('api/user/', payload)
+      expect(response).toEqual(payload)
     })
   })
 
-  describe('#update', function () {
-    it('patches a resource on the network', async function () {
+  describe('#update', () => {
+    it('patches a resource on the network', async () => {
       let payload = {foo: 'bar', fiz: {biz: 'buz'}}
-      server.respondWith('PATCH', '/api/user/2/', JSON.stringify(payload))
+      fetchMock.mock(/.*api\/user\/2\//g, payload, {
+        method: 'PATCH'
+      })
 
-      return expect(adapter.update('/api/user/2/', payload)).resolves.toEqual(payload)
+      const response = await adapter.update('api/user/2/', payload)
+      expect(response).toEqual(payload)
     })
   })
 
-  describe('#destroy', function () {
-    it('deletes a record from the network', function () {
-      server.respondWith('DELETE', '/api/user/42/', [204, {}, ''])
+  describe('#destroy', () => {
+    it('deletes a record from the network', async () => {
+      fetchMock.mock(/.*api\/user\/42\//g, 204, {
+        method: 'DELETE'
+      })
 
-      return adapter.destroy('/api/user/42/')
+      const response = await adapter.destroy('api/user/42/')
+      expect(response).toMatchObject({})
     })
   })
 
-  describe('#serializeRequests', function () {
-    it('parallel requests do not start when `serializeRequests = true`', async function () {
-      server.autoRespond = false
-      server.respondImmediately = false
+  describe('#serializeRequests', () => {
+    const path1 = 'api/user/42/'
+    const path2 = 'api/offer/9000/'
+
+    it('parallel requests do not start and start after previous finishes when `serializeRequests = true`', async () => {
+      adapter._makeRequest = jest.fn().mockResolvedValue({})
```

Comment:
I have concerns about this being equivalent to the previous tests; most of these tests rely on the promise returned by `_makeRequest` stalling until explicitly requested to resolve (hence `server.autoRespond = false` and `server.respondImmediately = false`. Have you tested that these tests do not pass if you remove the `adapter.serializeRequest = true` lines?

Reply:
I have. The tests will fail if serializeRequest is either commented out of set to false. The assertion `expect(adapter._makeRequest).toHaveBeenCalledTimes(1)` will fail because two promises will be in the outstanding requests queue, not one like in serializeRequest mode

Reply:
I'm having trouble understanding why that would be the case. If the promise returned by `_makeRequest` resolves immediately (i.e. the API request returns a value immediately), the second request's promise should look like: `Promise.all([Promise.resolve({}).catch(() => {})]).then(() => this._makeRequest(...))`. That should result in `_makeRequest` getting called another time.

Reply:
I changed the microtask ticks to `await new Promise(setImmediate)` (which will flush all microtask chains instead of just ticking once) and that test fails with multiple calls to `_makeRequest`. So I think what this test is testing is one microtask tick will only run to completion the `Promise.resolve({}).catch(() => {})` part above, not the `then(() => this._makeRequest(...))`. The tests really should have used `await new Promise(setImmediate)` from the beginning (that was my oversight) but in any case these rewritten tests are not accurately testing the desired IO behavior of this code.

Reply:
@JohnColanduoni That would make sense. If you flush the whole queue there will be multiple calls. I thought we wanted to test on the first tick on the microtask queue just to see the queued up state and retick it for future state. 

If we want to flush the whole queue from the beginning, we could essentially recreate some of the behavior from the sinon fakeServer with jest resolvers to test specific state, but that might be overkill. I would love to get your input/suggestions on this to take the best approach.

Reply:
What (that particular) test ultimately wants to know is in the situation where these promises are doing IO, that the second request will not start until the first one completes no matter how long that takes. The other tests are similar, in that they are interested in behavior on timescales irrelevant to particular microtask queue ticks.

I would suggest first replacing all the `await Promise.resolve().then(() => {})` with `await new Promise(setImmediate)` for these tests, so that they are categorically operating on that timescale. Then we can mimic the never resolved promise behavior of sinon with `.mockImplementation(() => new Promise(() => {}))` for the promises that were never resolved in the sinon version.

There are some tests that did trigger responses via the sinon `respond` method, but at specific times (e.g. the test named `sequential requests start after previous finishes when serializeRequests = true`). Those are tougher but can return manually resolving promises via something like:
```
let respondToRequest = []
adapter._makeRequest = jest.fn().mockImplementation(() => {
  return new Promise((resolve) => {
    respondToRequest.push(resolve)
  })
})
```
Then the equivalent of `server.requests[0].respond(200, { "Content-Type": "application/json" }, JSON.stringify(payload))`
is something like `respondToRequest[0](payload)`.

Reply:
@JohnColanduoni This is an awesome recommendation. It mocks the behavior we want while keeping it simple. I'll implement this soon and thanks for the patience and help with this!